site/content/docs/concepts/pipelines.en.md

```diff
@@ -146,12 +146,57 @@ After creating your pipeline, you can manage the version of Copilot used by your
+    pre_deployments:
+      db_migration: # The name of this action.
+        buildspec: copilot/pipelines/demo-api-frontend-main/buildspecs/buildspec.yml # The path to the buildspec.
+    deployments: # Optional, ordering of deployments. 
+      orders:
+      warehouse:
+      frontend:
+        depends_on: [orders, warehouse]
+    post_deployments:
+      db_migration:
+        buildspec: copilot/pipelines/demo-api-frontend-main/buildspecs/post_buildspec.yml
+      integration:
+        buildspec: copilot/pipelines/demo-api-frontend-main/buildspecs/integ-buildspec.yml
+        depends_on: [db_migration] # Optional, ordering of actions.
+```
+In the `buildspec` manifest field, add the path, relative to your project root, 
+of your [buildspec file]( https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html).
+The Copilot environment variables `COPILOT_APPLICATION_NAME` and `$COPILOT_ENVIRONMENT_NAME` are available
+for use within these buildspecs.
 
-Of course, one of the most important parts of a pipeline is the automated testing. To add tests, such as integration or end-to-end tests, that run after a deployment stage, include those commands in the `test_commands` section. If all the tests succeed, your change is promoted to the next stage.
+You may specify the run order of the actions using the `depends_on` subfield, just like you would to indicate your desired [order of deployments](#ordering).  
+
+!!! info
+    The CodeBuild projects generated for pre- and post-deployments and test commands are deployed in the same region as the pipeline and app. To access the VPC of the environment being deployed or deployed to, use Copilot commands in your pre-/post-deployment action buildspec or directly in your test commands.
 
-Adding `test_commands` generates a CodeBuild project with the [aws/codebuild/amazonlinux2-x86_64-standard:3.0](https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-available.html) image - so most commands from Amazon Linux 2 (including `make`) are available for use.
+### Ordering
+The `deployments` field enables you to specify the deployment order of 
+workloads or environments (depending on the type of pipeline). If this 
+is not specified, deployments run in parallel. (See [this blog post](../../blogs/release-v118.en.md#controlling-order-of-deployments-in-a-pipeline) for more info.)
 
+### Testing
+If your post-deployment testing requires only a handful of commands and doesn't necessarily
+warrant wiring up a separate buildspec, utilize the `test_commands` field.
+
+!!! warning
+    The `post_deployments` and `test_commands` fields within a stage are mutually exclusive.
+
+
+Pre-deployments, post-deployments, and test commands generate CodeBuild projects with the [aws/codebuild/amazonlinux2-x86_64-standard:3.0](https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-available.html) image, so most commands from Amazon Linux 2 (including `make`) are available for use. 
```

Comment:
Seems like we are actually using 4.0 🎊!  https://github.com/aws/copilot-cli/blob/d19b2af4e095db5a4366010bdc8e13beea0f810b/internal/pkg/deploy/pipeline.go#L36